File: packages/tui/src/theme.ts
import { visibleWidth } from '#tui/strings';
import { dim, yellow } from '#tui/theme/styles';
import type { Choice } from '#tui/types';

export const symbols = {
	question: '?',
	answer: '>',
	error: '!',
	info: 'i',
	selected: '●',
	unselected: '○',
	pointer: '›',
	success: '✓',
	warning: '!',
} as const;

export const renderQuestion = (message: string, hint?: string): string => {
	const suffix = hint ? ` ${hint}` : '';

	return `${symbols.question} ${message}${suffix} `;
};

export const renderError = (message: string): string => `${yellow(`  ⚠ ${message}`)}\n`;

export const renderChoices = <T>(choices: Array<Choice<T>>): string => {
	return choices
		.map((choice, index) => {
			const disabled = choice.disabled ? ` (${typeof choice.disabled === 'string' ? choice.disabled : 'disabled'})` : '';
			const hint = choice.hint ? ` ${choice.hint}` : '';

			return `  ${index + 1}. ${choice.label}${hint}${disabled}`;
		})
		.join('\n');
};

export const renderTable = (headers: string[], rows: string[][]): string => {
	const columnCount = Math.max(headers.length, ...rows.map((row) => row.length));

	if (!Number.isFinite(columnCount) || columnCount <= 0) {
		return '';
	}

	const widths = Array.from({ length: columnCount }, (_, index) => {
		return Math.max(visibleWidth(headers[index] ?? ''), ...rows.map((row) => visibleWidth(row[index] ?? '')));
	});

	const padVisible = (value: string, width: number): string => `${value}${' '.repeat(Math.max(0, width - visibleWidth(value)))}`;

	const renderRow = (columns: string[]): string => {
		return ` │ ${widths.map((width, index) => padVisible(columns[index] ?? '', width)).join(' │ ')} │`;
	};

	const renderBorder = (left: string, middle: string, right: string): string => {
		return ` ${left}${widths.map((width) => '─'.repeat(width + 2)).join(middle)}${right}`;
	};

	const top = renderBorder('┌', '┬', '┐');
	const divider = renderBorder('├', '┼', '┤');
	const bottom = renderBorder('└', '┴', '┘');

	if (headers.length === 0) {
		return [top, ...rows.map(renderRow), bottom].join('\n');
	}

	return [top, renderRow(headers.map(dim)), divider, ...rows.map(renderRow), bottom].join('\n');
};

File: console/shell/internal/theme/theme.go
package theme

import (
	"fmt"
	stdstrings "strings"

	text "github.com/oullin/tui/internal/strings"
)

const (
	SymbolQuestion   = "?"
	SymbolAnswer     = ">"
	SymbolError      = "!"
	SymbolInfo       = "i"
	SymbolSelected   = "●"
	SymbolUnselected = "○"
	SymbolPointer    = "›"
	SymbolSuccess    = "✓"
	SymbolWarning    = "!"
)

func RenderQuestion(message string, hint string) string {
	suffix := ""
	if hint != "" {
		suffix = " " + hint
	}

	return fmt.Sprintf("%s %s%s ", SymbolQuestion, message, suffix)
}

func RenderError(message string) string {
	return Red(fmt.Sprintf("%s %s", SymbolError, message)) + "\n"
}

func RenderTable(headers []string, rows [][]string) string {
	columnCount := len(headers)
	for _, row := range rows {
		if len(row) > columnCount {
			columnCount = len(row)
		}
	}

	if columnCount <= 0 {
		return ""
	}

	widths := make([]int, columnCount)
	for index := range columnCount {
		widths[index] = text.VisibleWidth(valueAt(headers, index))
		for _, row := range rows {
			width := text.VisibleWidth(valueAt(row, index))
			if width > widths[index] {
				widths[index] = width
			}
		}
	}

	renderRow := func(columns []string) string {
		parts := make([]string, len(widths))
		for index, width := range widths {
			parts[index] = padVisible(valueAt(columns, index), width)
		}

		return " │ " + stdstrings.Join(parts, " │ ") + " │"
	}

	renderBorder := func(left string, middle string, right string) string {
		parts := make([]string, len(widths))
		for index, width := range widths {
			parts[index] = stdstrings.Repeat("─", width+2)
		}

		return " " + left + stdstrings.Join(parts, middle) + right
	}

	top := renderBorder("┌", "┬", "┐")
	divider := renderBorder("├", "┼", "┤")
	bottom := renderBorder("└", "┴", "┘")

	if len(headers) == 0 {
		lines := append([]string{top}, mapRows(rows, renderRow)...)
		lines = append(lines, bottom)
		return stdstrings.Join(lines, "\n")
	}

	dimmedHeaders := make([]string, len(headers))
	for index, header := range headers {
		dimmedHeaders[index] = Dim(header)
	}

	lines := []string{top, renderRow(dimmedHeaders), divider}
	lines = append(lines, mapRows(rows, renderRow)...)
	lines = append(lines, bottom)
	return stdstrings.Join(lines, "\n")
}

func valueAt(values []string, index int) string {
	if index < 0 || index >= len(values) {
		return ""
	}

	return values[index]
}

func padVisible(value string, width int) string {
	padding := width - text.VisibleWidth(value)
	if padding < 0 {
		padding = 0
	}

	return value + stdstrings.Repeat(" ", padding)
}

func mapRows(rows [][]string, render func([]string) string) []string {
	result := make([]string, len(rows))
	for index, row := range rows {
		result[index] = render(row)
	}

	return result
}
